azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -60,14 +61,21 @@ def __init__(self, wire_client, full_goal_state=False, base_incarnation=None):
         directly.
 
         """
-        try:
-            uri = GOAL_STATE_URI.format(wire_client.get_endpoint())
+        uri = GOAL_STATE_URI.format(wire_client.get_endpoint())
+
+        for _ in range(1, 6):
             self.xml_text = wire_client.fetch_config(uri, wire_client.get_header())
             xml_doc = parse_doc(self.xml_text)
 
+            role_instance = find(xml_doc, "RoleInstance")
+            if role_instance:
+                break
+        else:
+            raise IncompleteGoalStateError("Goal State doesn't have a RoleInstance.")
```

Comment:
we need to log the incarnation somewhere

Reply:
(at least for the extension handler)

Reply:
Logging here has the potential of being very verbose  since this code is called in each iteration of the main loop. From what we've seen so far this error does not seem to last for long, so likely we are ok as is.

Sorry, for my previous comment, I mean logging in the extension handler code (see try_update_goal_state). The verbosity issue is already handled in that code. Logging the incarnation number could be accomplished by simply adding it to the message in the exception (the code in try_update_goal_state logs only the first error, so if a different kind of error happens we wouldn't know until the operation succeeds and the error state is reset, but that is a different issue.)

If we decide to leave the logging here, let's change the error message to something like

> Fetched goal state without a RoleInstance [incarnation {inc}]".format(inc=self.incarnation))